Project2015To2017/Reading/Conditionals/CharacterUtilities.cs

```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Project2015To2017.Reading.Conditionals
+{
+    internal static class CharacterUtilities
+    {
+        static internal bool IsNumberStart(char candidate)
+        {
+            return (candidate == '+' || candidate == '-' || candidate == '.' || char.IsDigit(candidate));
+        }
+
+        static internal bool IsSimpleStringStart(char candidate)
+        {
+            return (candidate == '_' || char.IsLetter(candidate));
+        }
+
+        static internal bool IsSimpleStringChar(char candidate)
+        {
+            return (IsSimpleStringStart(candidate) || char.IsDigit(candidate));
+        }
+
+        static internal bool IsHexAlphabetic(char candidate)
+        {
+            return (candidate == 'a' || candidate == 'b' || candidate == 'c' || candidate == 'd' || candidate == 'e' || candidate == 'f' ||
```

Comment:
return (candidate >= 97 && candidate <= 102) || (candidate >= 65 && candidate <= 70)
would work too and is probably a lot faster

Reply:
Well, that's MSBuild. :) I really don't think we should bring in our own modifications except those to reduce dependencies on MSBuild code. We might later want to update code from upstream so such changes might introduce confusion.

Reply:
Interesting, can we include a link to the source somewhere here in case things get outdated?

Reply:
Introduced upstream.md in *Reading\* directory.

Reply:
I think the answer is no, but would https://github.com/daveaglick/Buildalyzer be of any use here to avoid copying the code in? It would be a bit too hefty a dependency probably and may not actually help. Or are git sub-modules another possibility to directly bring in the code from MS Build? I've never actually used them before.

Reply:
@mungojam unfortunately, we cannot bring in the code directly. It requires removal of specific code, non-trivial replacements in a couple of places so that there are no dependencies introduced. And since we don't want our code to bring in the whole enormous dependency on MSBuild *and* its dependencies *and* .NET Core SDK *and* Microsoft SDKs *and* NuGet, Buildalyzer (that does exactly the thing) can be of no use to us.